### PR TITLE
[WIP] Experimental GPU Support

### PR DESCRIPTION
**Pull request by wavesoft**
- Enabling GPU support on marathon
- Adding general purpose isolator detection to private/public slaves in mesos package (currently only detecting GPUs and adding `gpu/nvidia` isolator)

### CODE REVIEW

**Comment by cmaloney** [comment]
@wavesoft any timeline on the work going to happen on making this landable as previously discussed?

**Comment by wavesoft** [comment]
Hey @cmaloney ! I will have a look now.

File: packages/mesos/extra/make_isolators.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    nvidia_smi_binary = '/bin/nvidia-smi'
+
+    # Make sure file is there and with correct permissions
+    if not os.path.isfile(nvidia_smi_binary):
+        return False
+    if not os.access(nvidia_smi_binary, os.X_OK):
+        return False
+
+    # Try to run it
```
Comment (orsenthil) [inline packages/mesos/extra/make_isolators.py]: [readability review] - please kill all the comments in this file. In practice, I have seen comments that explain "why" (when needed) are helpful.  Comments that explain "what" should be avoided.

File: packages/mesos/build [hunk 2]
```diff
@@ -64,6 +64,22 @@ disk_resource_script="$PKG_PATH/bin/make_disk_resources.py"
...
+# (such as GPU) and creates an EnvironmentFile that contains a MESOS_ISOLATION env variable, with
+# additional isolators APPENDED to the current environment.
+
+public_isolators_service="$PKG_PATH"/dcos.target.wants_slave_public/dcos-isolators-discovery-pub-agent.service
+mkdir -p "$(dirname ${public_isolators_service})"
+envsubst '$PKG_PATH' < /pkg/extra/dcos-isolators-discovery-pub-agent.service > "$public_isolators_service"
+
+private_isolators_service="$PKG_PATH"/dcos.target.wants_slave/dcos-isolators-discovery-priv-agent.service
+mkdir -p "$(dirname ${private_isolators_service})"
+envsubst '$PKG_PATH' < /pkg/extra/dcos-isolators-discovery-priv-agent.service > "$private_isolators_service"
```
Comment (wavesoft) [inline packages/mesos/build]: Replying to @kaysoky 's comment in a previous PR:
> Do you really need a new service for this?
>
> Seems more appropriate to add a ExecStartPre=... to the two agent services.

Correct me if I am wrong, I don't have a linux installation to quickly test this, but aren't the `EnvironmentFile`s loaded before the `ExecStartPre` / `ExecStart` ? If true, the `ExecStartPre` won't be able to generate the environment variables on time...

File: packages/mesos/extra/make_isolators.py (hunk 1 above)
Comment (wavesoft) [inline packages/mesos/extra/make_isolators.py]: Will do!

**Comment by cmaloney** [comment]
To note some of the more core problems here
1. We're writing a giant program to directly reverse-engineer / undo a bunch of logic mesos is doing internally. Duplicating code to get the behavior we want effectively. Need to unwind it so mesos has a reasonable option we can use. That could be something like "when there are no gpus but gpu isolator is specified it gets skipped".
2. We don't want to run another separate discovery service. These make it incredibly hard to manage hosts and change resources / attributes / etc. Things have to go inline in the mesos slave service, and we definitely don't want to add more python scripts here.

General thing: If to land something in the product we have to write a bunch of code to undo engineering decisions made by engineers employed by mesosphere and make things work reasonably for our use case, there's a problem which needs to be fixed. The baseline components need to be as easy to use as possible. If we have to build this much extra tooling around mesos to use a feature, all our customers probably have to do it as well.

**Comment by mgummelt** [comment]
@wavesoft I'd like to build this cluster to test Spark GPU support.  Do you have time to make it so the build passes so I can do that?

**Comment by wavesoft** [comment]
@mgummelt will work on it later today!

**Comment by wavesoft** [comment]
Hi all, I based this PR over `dcos/1.8.1` branch and the errors listed does not seem to come from my additions. Any suggestions? Shall I try to rebase on `master` ?

**Comment by cmaloney** [comment]
We're adding short term experimental gpu support via #766 (all the cluster has GPUs or none of them)

**Pull request closed without merging**